Merge sdk and catalog docs into kpt

prachirp opened the pull request:
Combine all documentation from kpt-functions-sdk and kpt-functions-catalog with existing kpt docs. Main changes:

- Reorder homepage to move install section to top, add search bar
- Add from kpt-functions-sdk: concepts/functions, guides/consumer/function/pipeline, guides/producer/functions/ts, guides/producer/functions/ts/develop, layout, and static/images
- Add from kpt-functions-catalog: guides/consumer/function/catalog
- Create next steps section at the bottom of all kpt function docs
- Write influences section: concepts/architecture
- Write declarative call section: guides/consumer/function
- Write structured results section: reference/fn/run

Fixes GoogleContainerTools/kpt#640

frankfarzan commented on site/content/en/guides/consumer/function/catalog/_index.md:
```diff
@@ -0,0 +1,76 @@
+---
+
+<!---
+DO NOT EDIT. Generated by: "cd catalog; npm run gen-docs"
```
Where's the package after this move? At very least fully specify the path, but ideally this should be part of a top-level command when generating all docs.

prachirp replied:
The catalog package is at googlecontainertools.github.io/kpt/guides/consumer/function/catalog. I also included a list of where sdk docs ended up in the description.

prachirp replied:
Done

frankfarzan commented on site/content/en/guides/producer/functions/ts/_index.md:
```diff
@@ -4,8 +4,103 @@ linkTitle: "TypeScript Function SDK"
+   export CONFIGS=../../example-configs
+
+   kpt fn source $CONFIGS |
+   node dist/label_namespace_run.js -d label_name=color -d label_value=orange |
```
we should show exec method of running with node.

prachirp replied:
What do you mean by this?

prachirp replied:
@frankfarzan I filed https://github.com/GoogleContainerTools/kpt/issues/768 since I wasn't able to get the exec method working. Let's fix this in a separate PR.

frankfarzan commented on site/content/en/guides/producer/functions/ts/_index.md:
````diff
@@ -4,8 +4,103 @@ linkTitle: "TypeScript Function SDK"
+   npm run watch
+   ```
+
+1. Run the `label_namespace` function:
````
In this file we should include the function inline (I think there's a way to link instead of copy code). This gives the reader an idea of what TS functions look like and is consistent with other languages.

prachirp replied:
Linked to the function.

frankfarzan commented on site/content/en/guides/producer/functions/ts/_index.md:
```diff
@@ -4,8 +4,103 @@ linkTitle: "TypeScript Function SDK"
+1. Try modifying the function in `src/label_namespace.ts` to perform other operations
+   on `example-configs`, then repeat step 4.
+
+   The function should implement the `KptFunc` interface [documented here][api-kptfunc].
```
Similarly, we can include other interesting functions.

prachirp replied:
Added suggest psp and validate rolebinding as examples of other interesting functions.

frankfarzan commented on site/content/en/guides/producer/functions/ts/develop/_index.md:
```diff
@@ -0,0 +1,253 @@
+[install-node]: https://github.com/nodejs/help/wiki/Installation
+[install-docker]: https://docs.docker.com/v17.09/engine/installation
+[beta-feature]: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.15.md#customresourcedefinition-openapi-publishing
+[run functions]: ../../../../consumer/function/
```
Can you double check there are no broken links (Changes are high with such moves). Maybe also look into whether there's an automated way to catch broken links.

prachirp replied:
I installed and used muffet locally. https://github.com/raviqqe/muffet
I also ran the W3 link checker on the existing kpt, kpt-functions-sdk, and kpt-functions-catalog websites and fixed some of the links it caught.

prachirp commented:
cc @phanimarupaka Can you take a look at these changes too? Mostly need review of pages unrelated to functions:
- site/content/en/_index.md
- site/content/en/concepts/architecture/_index.md

phanimarupaka commented on site/content/en/_index.md:
```diff
@@ -16,6 +16,17 @@ apply Kubernetes configuration.
 
 {{% /blocks/lead %}}
 
+{{< blocks/section color="light" >}}
```
What is the motivation behind this change? Can you attach a screenshot as it how it looks like after the change? Do you think the installation icon is not at right place?

prachirp replied:
Having the installation link clearly at the top makes it easier for newcomers to take the first step and install kpt. It also brings us more in line with other websites like docs.docker.com, nodejs.org, helm.sh which all have installation directions first, then guides/reference. Also, the install icon was broken and I fixed it.
Before: 
<img width="295" alt="Screen Shot 2020-06-29 at 11 13 34 AM" src="https://user-images.githubusercontent.com/2899196/86023948-39fec700-b9e1-11ea-97ba-a0b28960e9e2.png">
After:
<img width="283" alt="Screen Shot 2020-06-29 at 11 14 30 AM" src="https://user-images.githubusercontent.com/2899196/86023966-3f5c1180-b9e1-11ea-8fe0-d1cb21ff35d3.png">

prachirp commented:
@phanimarupaka Let's merge this. We can address other comments to the site's look-and-feel in followup PRs.

frankfarzan commented:
Some follow up comments:
- Auto-generated API(s) documetnation (typedoc) is broken now
- Separate Quick Guide into a separate page
- Hugo should have a way to include source functions directly
- We deleted the "how to run functions" page from SDK, was there anything we wanted to carry over?

prachirp commented:
> Some follow up comments:
> 
> * Auto-generated API(s) documetnation (typedoc) is broken now
Fixing typedoc.
> * Separate Quick Guide into a separate page
Separating Quickstart.
> * Hugo should have a way to include source functions directly
Hugo allows shortcodes so I added one to include source code.
> * We deleted the "how to run functions" page from SDK, was there anything we wanted to carry over?
No, I split the page sections between package publisher and consumer functions docs.
